internal/file/proto_set_provider.go

```diff
@@ -0,0 +1,280 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package file
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/uber/prototool/internal/x/settings"
+	"go.uber.org/zap"
+)
+
+type protoSetProvider struct {
+	logger         *zap.Logger
+	walkTimeout    time.Duration
+	configProvider settings.ConfigProvider
+}
+
+func newProtoSetProvider(options ...ProtoSetProviderOption) *protoSetProvider {
+	protoSetProvider := &protoSetProvider{
+		logger:      zap.NewNop(),
+		walkTimeout: DefaultWalkTimeout,
+	}
+	for _, option := range options {
+		option(protoSetProvider)
+	}
+	protoSetProvider.configProvider = settings.NewConfigProvider(
+		settings.ConfigProviderWithLogger(protoSetProvider.logger),
+	)
+	return protoSetProvider
+}
+
+func (c *protoSetProvider) GetForDir(workDirPath string, dirPath string) ([]*ProtoSet, error) {
+	workDirPath, err := absClean(workDirPath)
+	if err != nil {
+		return nil, err
+	}
+	absDirPath, err := absClean(dirPath)
+	if err != nil {
+		return nil, err
+	}
+	configFilePath, err := c.configProvider.GetFilePathForDir(absDirPath)
+	if err != nil {
+		return nil, err
+	}
+	// we need everything for generation, not just the files in the given directory
+	// so we go back to the config file if it is shallower
+	// display path will be unaffected as this is based on workDirPath
+	configDirPath := absDirPath
+	if configFilePath != "" {
+		configDirPath = filepath.Dir(configFilePath)
+	}
+
+	protoFiles, err := c.walkAndGetAllProtoFiles(workDirPath, configDirPath)
+	if err != nil {
+		return nil, err
+	}
+	dirPathToProtoFiles := getDirPathToProtoFiles(protoFiles)
+	protoSets, err := c.getBaseProtoSets(dirPathToProtoFiles)
+	if err != nil {
+		return nil, err
+	}
+	for _, protoSet := range protoSets {
+		protoSet.WorkDirPath = workDirPath
+		protoSet.DirPath = absDirPath
+	}
+	c.logger.Debug("returning ProtoSets", zap.String("workDirPath", workDirPath), zap.String("dirPath", dirPath), zap.Any("protoSets", protoSets))
+	return protoSets, nil
+}
+
+func (c *protoSetProvider) GetForFiles(workDirPath string, filePaths ...string) ([]*ProtoSet, error) {
+	workDirPath, err := absClean(workDirPath)
+	if err != nil {
+		return nil, err
+	}
+	protoFiles, err := getProtoFiles(filePaths)
+	if err != nil {
+		return nil, err
+	}
+	dirPathToProtoFiles := getDirPathToProtoFiles(protoFiles)
+	protoSets, err := c.getBaseProtoSets(dirPathToProtoFiles)
+	if err != nil {
+		return nil, err
+	}
+	for _, protoSet := range protoSets {
+		protoSet.WorkDirPath = workDirPath
+		protoSet.DirPath = workDirPath
+	}
+	c.logger.Debug("returning ProtoSets", zap.String("workDirPath", workDirPath), zap.Strings("filePaths", filePaths), zap.Any("protoSets", protoSets))
+	return protoSets, nil
+}
+
+func (c *protoSetProvider) getBaseProtoSets(dirPathToProtoFiles map[string][]*ProtoFile) ([]*ProtoSet, error) {
+	// TODO: this should be handled elsewhere
```

Comment:
Seems right to me. I generally exit out of a function early in this scenario just as a safety measure/as a sanity check, so that the remainder of the logic in the function doesn't need to handle this case, but I removed this.